drivers/can/can_nxp_s32_canxl.c

```diff
@@ -0,0 +1,977 @@
+	!(DT_INST_NODE_HAS_PROP(inst, prop_seg_data) &&		\
+	DT_INST_NODE_HAS_PROP(inst, phase_seg1_data) &&		\
+	DT_INST_NODE_HAS_PROP(inst, phase_seg2_data))) ||
+
+#if DT_INST_FOREACH_STATUS_OKAY(SP_AND_TIMING_DATA_NOT_SET) 0
+#error You must either set a sampling-point-data or timings (phase-seg-data* and prop-seg-data)
+#endif
+#endif
+
+
+static int can_nxp_s32_get_capabilities(const struct device *dev, can_mode_t *cap)
+{
+	ARG_UNUSED(dev);
+
+	*cap = CAN_MODE_NORMAL | CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY | CAN_MODE_FD;
+
+	return 0;
+}
+
+static int can_nxp_s32_start(const struct device *dev)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+	struct can_nxp_s32_data *data = dev->data;
+	int err;
+
+	if (data->started) {
+		return -EALREADY;
+	}
+
+	Canexcel_Ip_EnterFreezeMode(config->instance);
+
+	if (config->phy != NULL) {
+		err = can_transceiver_enable(config->phy);
+		if (err != 0) {
+			LOG_ERR("failed to enable CAN transceiver (err %d)", err);
+			return err;
+		}
+	}
+
+	Canexcel_Ip_ExitFreezeMode(config->instance);
+
+	data->started = true;
+
+	return 0;
+}
+
+static int can_nxp_s32_abort_msg(const struct can_nxp_s32_config *config, int mb_idx)
+{
+	uint32_t time_start = 0;
+	int ret = 0;
+
+	Canexcel_Ip_EnterFreezeMode(config->instance);
+
+	CanXL_ClearMsgBuffIntCmd(config->base_grp_ctrl, mb_idx);
+	CanXL_ClearMsgDescIntStatusFlag(config->base_grp_ctrl, mb_idx);
+
+	time_start = k_uptime_get();
+	/* Set system lock Status */
+	(void)config->base_dsc_ctrl->DSCMBCTRLAR[mb_idx].SYSLOCK.DCSYSLOCK;
+	while (CanXL_GetDescControlStatus(config->base_dsc_ctrl, mb_idx)
+			== CANEXCEL_DESCNTSTATUS_LOCKED_HW) {
+		if (k_uptime_get() - time_start >= CAN_NXP_S32_TIMEOUT_MS) {
+			ret = CANEXCEL_STATUS_TIMEOUT;
+			break;
+		}
+	}
+
+	/* Inactive descriptor */
+	config->base_dsc_ctrl->DSCMBCTRLAR[mb_idx].ACT.DCACT = 0;
+
+	Canexcel_Ip_ExitFreezeMode(config->instance);
+
+	return ret;
+}
+
+static int can_nxp_s32_stop(const struct device *dev)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+	struct can_nxp_s32_data *data = dev->data;
+	can_tx_callback_t function;
+	void *arg;
+	int alloc;
+	int err;
+
+	if (!data->started) {
+		return -EALREADY;
+	}
+
+	data->started = false;
+
+	/* Abort any pending TX frames before entering freeze mode */
+	for (alloc = 0; alloc < CAN_NXP_S32_MAX_TX; alloc++) {
+		function = data->tx_cbs[alloc].function;
+		arg = data->tx_cbs[alloc].arg;
+
+		if (atomic_test_and_clear_bit(data->tx_allocs, alloc)) {
+			if (can_nxp_s32_abort_msg(config,
+					ALLOC_IDX_TO_TXMB_IDX(alloc))) {
+				LOG_ERR("Can't abort message !");
+			};
+
+			function(dev, -ENETDOWN, arg);
+			k_sem_give(&data->tx_allocs_sem);
+		}
+	}
+
+	Canexcel_Ip_EnterFreezeMode(config->instance);
+
+	if (config->phy != NULL) {
+		err = can_transceiver_disable(config->phy);
+		if (err != 0) {
+			LOG_ERR("failed to disable CAN transceiver (err %d)", err);
+			return err;
+		}
+	}
+
+	Canexcel_Ip_ExitFreezeMode(config->instance);
```

Comment:
Is it necessary to enter freeze mode in order to disable the PHY? If so, why?

This function is supposed to stop the CAN controller so that it does not send ACKs for any frames on the bus. I am unable to tell if this is done here?

Reply:
yes, really no necessary. I removed